### DISCUSSION TITLE
v1.7.3-pre

### DISCUSSION THREAD

**Discussion by ggerganov** [ggml-org/whisper.cpp#2616] [Announcements]
## Overview

Massive performance improvements for the Metal backend, especially for beams > 1. Especially for quantized models.
Setting as "pre-release" since there have been major changes to the build system (now using CMake) and I wan't to gather some feedback about how well the project builds now on various platforms. Please leave comments in the discussion to help fix any remaining issues before the official release.

|    CPU |  Config |         Model |  Th |  FA |    Enc. |    Dec. |    Bch5 |      PP |  Commit |
|    --- |     --- |           --- | --- | --- |     --- |     --- |     --- |     --- |     --- |
| M2 Ultra | Metal |          tiny |   1 |   1 |    7.90 |    1.26 |    0.35 |    0.01 | ed733e8 |
| M2 Ultra | Metal |     tiny-q5_0 |   1 |   1 |    8.44 |    1.23 |    0.36 |    0.01 | ed733e8 |
| M2 Ultra | Metal |     tiny-q5_1 |   1 |   1 |    8.26 |    1.27 |    0.37 |    0.01 | ed733e8 |
| M2 Ultra | Metal |     tiny-q8_0 |   1 |   1 |    8.03 |    1.21 |    0.35 |    0.01 | ed733e8 |
| M2 Ultra | Metal |          base |   1 |   1 |   13.77 |    1.80 |    0.42 |    0.02 | ed733e8 |
| M2 Ultra | Metal |     base-q5_0 |   1 |   1 |   15.02 |    1.72 |    0.42 |    0.02 | ed733e8 |
| M2 Ultra | Metal |     base-q5_1 |   1 |   1 |   14.93 |    1.74 |    0.42 |    0.02 | ed733e8 |
| M2 Ultra | Metal |     base-q8_0 |   1 |   1 |   14.26 |    1.68 |    0.41 |    0.02 | ed733e8 |
| M2 Ultra | Metal |         small |   1 |   1 |   39.76 |    3.54 |    0.85 |    0.05 | ed733e8 |
| M2 Ultra | Metal |    small-q5_0 |   1 |   1 |   45.07 |    3.47 |    0.87 |    0.05 | ed733e8 |
| M2 Ultra | Metal |    small-q5_1 |   1 |   1 |   44.82 |    3.49 |    0.87 |    0.05 | ed733e8 |
| M2 Ultra | Metal |    small-q8_0 |   1 |   1 |   41.79 |    3.30 |    0.84 |    0.05 | ed733e8 |
| M2 Ultra | Metal |        medium |   1 |   1 |  106.73 |    7.28 |    1.78 |    0.11 | ed733e8 |
| M2 Ultra | Metal |   medium-q5_0 |   1 |   1 |  124.43 |    6.63 |    1.83 |    0.12 | ed733e8 |
| M2 Ultra | Metal |   medium-q5_1 |   1 |   1 |  124.19 |    6.70 |    1.84 |    0.12 | ed733e8 |
| M2 Ultra | Metal |   medium-q8_0 |   1 |   1 |  113.88 |    6.52 |    1.75 |    0.11 | ed733e8 |
| M2 Ultra | Metal |    medium-dis |   1 |   1 |   94.97 |    0.97 |    0.22 |    0.01 | ed733e8 |
| M2 Ultra | Metal |      large-v2 |   1 |   1 |  193.33 |   10.53 |    2.65 |    0.20 | ed733e8 |
| M2 Ultra | Metal | large-v2-q5_0 |   1 |   1 |  229.22 |    9.52 |    2.72 |    0.23 | ed733e8 |
| M2 Ultra | Metal | large-v2-q5_1 |   1 |   1 |  229.40 |    9.62 |    2.73 |    0.23 | ed733e8 |
| M2 Ultra | Metal | large-v2-q8_0 |   1 |   1 |  207.30 |    9.36 |    2.59 |    0.21 | ed733e8 |
| M2 Ultra | Metal |  large-v2-dis |   1 |   1 |  171.43 |    1.09 |    0.25 |    0.02 | ed733e8 |
| M2 Ultra | Metal | large-v3-turbo |   1 |   1 |  173.45 |    1.73 |    0.41 |    0.03 | ed733e8 |
| M2 Ultra | Metal | large-v3-turbo-q5_0 |   1 |   1 |  205.52 |    1.52 |    0.42 |    0.04 | ed733e8 |
| M2 Ultra | Metal | large-v3-turbo-q8_0 |   1 |   1 |  185.90 |    1.48 |    0.40 |    0.03 | ed733e8 |

## What's Changed
* sync : ggml by @ggerganov in https://github.com/ggerganov/whisper.cpp/pull/2573
* ruby : Follow source tree change by @KitaitiMakoto in https://github.com/ggerganov/whisper.cpp/pull/2580
* Add `q8_0` models to `download-ggml-model.sh` by @mrienstra in https://github.com/ggerganov/whisper.cpp/pull/2589
* ruby : Add low-level methods to transcribe by @KitaitiMakoto in https://github.com/ggerganov/whisper.cpp/pull/2585
* sync : ggml by @ggerganov in https://github.com/ggerganov/whisper.cpp/pull/2608

**Full Changelog**: https://github.com/ggerganov/whisper.cpp/compare/v1.7.2...v1.7.3-pre

<hr /><em>This discussion was created from the release <a href='https://github.com/ggerganov/whisper.cpp/releases/tag/v1.7.3-pre'>v1.7.3-pre</a>.</em>

**Comment by zapsleep** [comment]
Xcode 16.1 and 16.2RC:
Swift Package Manager gives a warning `couldn't find pc file for whisper` when updating the package to ed733e85a1b399e215ab84cff91abe25f5f79d76.
Then when trying to Build, it fails with:
`'whisper.h' file not found with <angled> include; use "quotes" instead`
in whisper/Sources/whisper/whisper.h
and every file with `import whisper` doesn't see imported functions and types.

**Reply by ggerganov** [reply]
The SPM package now requires to build the `whisper.cpp` binaries separately using CMake. The Github workflow demonstrates one way to do that:

https://github.com/ggerganov/whisper.cpp/blob/6aa1d7b89216dc6d9c41fc1536fae59adb214024/.github/workflows/build.yml#L522-L566

This is one of the main changes in the build process for this release. It would be great if we get some help with providing build instructions that are easy to follow and work for various deployment targets. For more info, also check https://github.com/ggerganov/llama.cpp/issues/10747#issuecomment-2538203367

**Comment by mrfragger** [comment]
-- Found Git: /usr/bin/git (found version "2.39.3 (Apple Git-146)")
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Warning: ccache not found - consider installing it for faster compilation or disable this warning with GGML_CCACHE=OFF
-- CMAKE_SYSTEM_PROCESSOR: arm64
-- Including CPU backend
-- Accelerate framework found
-- Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES) 
-- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES) 
-- Could NOT find OpenMP (missing: OpenMP_C_FOUND OpenMP_CXX_FOUND) 
CMake Warning at ggml/src/ggml-cpu/CMakeLists.txt:49 (message):
  OpenMP not found
Call Stack (most recent call first):
  ggml/src/CMakeLists.txt:305 (ggml_add_cpu_backend_variant_impl)


brew install libomp
and new ternimal tab bu still same error.
Must one download https://mac.r-project.org/openmp/
with curl perhaps?

edit:

ok on Mac OS 14.5
curl -O https://mac.r-project.org/openmp/openmp-15.0.7-darwin20-Release.tar.gz
 sudo tar fvxz openmp-15.0.7-darwin20-Release.tar.gz -C /

so that did the trick
-- Found OpenMP_C: -Xclang -fopenmp (found version "5.0")
-- Found OpenMP_CXX: -Xclang -fopenmp (found version "5.0")
-- Found OpenMP: TRUE (found version "5.0")

but stuck on git error I'll have to fix somehow
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: /usr/bin/git (found version "2.39.3 (Apple Git-146)")
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success

edit 2:
git init (someone suggested but got back)
-- Found Git: /usr/bin/git (found version "2.39.3 (Apple Git-146)")
fatal: Needed a single revision
fatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:

gonna stick with 1.7.2 for now though as it's working fine. Just occasionally freezes here and there but only about 5% of the time.

**Comment by mrfragger** [comment]
[ 86%] Building CXX object examples/server/CMakeFiles/server.dir/server.cpp.o
/Users/username/whisper/whisper.cpp-1.7.3-pre/examples/server/server.cpp:680:57: warning: 'tmpnam' is deprecated: This function is provided for compatibility reasons only.  Due to security concerns inherent in the design of tmpnam(3), it is highly recommended that you use mkstemp(3) instead. [-Wdeprecated-declarations]
            const std::string temp_filename_base = std::tmpnam(nullptr);
                                                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/stdio.h:189:1: note: 'tmpnam' has been explicitly marked deprecated here
__deprecated_msg("This function is provided for compatibility reasons only.  Due to security concerns inherent in the design of tmpnam(3), it is highly recommended that you use mkstemp(3) instead.")
^
/Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/include/sys/cdefs.h:218:48: note: expanded from macro '__deprecated_msg'
        #define __deprecated_msg(_msg) __attribute__((__deprecated__(_msg)))
                                                      ^
1 warning generated.
[ 89%] Linking CXX executable ../../bin/server
[ 89%] Built target server
[ 91%] Building CXX object examples/bench/CMakeFiles/bench.dir/bench.cpp.o
[ 94%] Linking CXX executable ../../bin/bench
[ 94%] Built target bench
[ 97%] Building CXX object examples/quantize/CMakeFiles/quantize.dir/quantize.cpp.o

**Comment by itsthisjustin** [comment]
I'm having all kinds of issues getting this working in my project but even after following the build commands here I get this error in the example project:

Building for 'iOS', but linking in dylib (/usr/local/lib/libggml.dylib) built for 'macOS'

**Reply by zapsleep** [reply]
I was having the same issue at first, cmake generated 'whisper.cpp.xcodeproj' configured for macOS for some reason. Maybe I wasn't careful with cmake flags.
It seems, after the xcodeproj is generated, calling `cmake -G Xcode ..` doesn't regenerate it even with new parameters. Maybe I'm wrong, but in my case after it was generated with macOS, it doesn't change to iOS with `cmake -G Xcode <...> -D CMAKE_SYSTEM_NAME=iOS <...>`.
What helped was removing 'whisper.cpp.xcodeproj' generated by `cmake` (I remove the whole `/build` dir completely) and generate again. But this time ensure you pass `-D CMAKE_SYSTEM_NAME=iOS`. For me it works every time: remove and then generate with 'iOS' or 'Darwin' (for macOS).
Then build and install again.

**Reply by itsthisjustin** [reply]
Any advice on how to do this for a project that has both iOS and macOS builds in it? I'd hate to have to do this every single time I want to build for one or the other

**Reply by zapsleep** [reply]
Not much, unfortunately ):
As I understand it, you just build dylibs once and then just link to them. No need to regenerate on every build.
But now the problems are:
1. Dynamic libraries need to be shipped with the app. Installing them into `/usr/local/lib` wouldn't do. And iOS apps (not sure about macOS) have specific requirements about dylibs locations inside the bundle. I had my share of trouble with `@rpath` evaluation for that reason. (today's XCode doesn't look in `/usr/local/lib` anymore as I understand)
I got the impression that Apple insists on packing dylibs as Frameworks and including them into bundles like that. It removes the trouble with runtime search paths completely. You can add 'Copy Files' phase to your XCode target, choose 'Framework' destination and add all the dylibs there.
2. We now have dylibs for iOS, iOS simulator and macOS (maybe even Catalyst). I guess they should be linked conditionally depending on target platform. Another way would be making fat libraries.

It becomes cumbersome. But this all is just a state of my understanding. I may be wrong and overcomplicating, and eventually someone with better knowledge of linking dylibs in XCode comes in and show an obvious easy way (:

**Comment by bingcheng1998** [comment]
follow those step and you can build success:
https://github.com/ggerganov/whisper.cpp/issues/2646#issuecomment-2690499223
